c_src/erl_optics.c

```diff
@@ -46,14 +48,48 @@ static int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info) {
     atom_dist = enif_make_atom(env, "dist");
     atom_histo = enif_make_atom(env, "histo");
     atom_quantile = enif_make_atom(env, "quantile");
-    poller = optics_poller_alloc();
-    ctx = enif_alloc(sizeof(struct map_ctx));
-    if (poller) {
-      optics_poller_backend(poller, (void *) ctx, backend_eo, backend_free_eo);
-    }
     return 0;
 }
 
+static ERL_NIF_TERM eo_register_carbon_poller(
+    ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+    if (!poller){
+        poller = optics_poller_alloc();
+    }
+    if (!poller) return ERROR("poller_allocation_error");
+    if (carbon_poller) return ERROR("carbon_poller_already_registered");
+
+    enum {host_size = 255, port_size = 255};
+    char host[host_size] = {0};
+    char port[port_size] = {0};
+    if (enif_get_string(env, argv[1], host, host_size, ERL_NIF_LATIN1) < 1 ||
+        enif_get_string(env, argv[2], port, port_size, ERL_NIF_LATIN1) < 1) {
+        return ERROR("unable_to_decode(hostname|port)");
+    }
+    optics_dump_carbon(poller, host, port);
+    carbon_poller = true;
+    return atom_ok;
+}
+
+static ERL_NIF_TERM eo_register_erlang_poller(
+    ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+    if (!poller){
+        poller = optics_poller_alloc();
+    }
+    if (!poller) return ERROR("poller_allocation_error");
+
+    if (!erlang_poller) {
+        ctx = enif_alloc(sizeof(struct map_ctx));
+        optics_poller_backend(poller, ctx, backend_eo, backend_free_eo);
+        erlang_poller = true;
+    }
+    else return ERROR("erlang_poller_aldready_registered");
```

Comment:
Very minor, but I would suggest structuring your code like this:
```
    if (!poller) return ERROR("poller_allocation_error");
    if (erlang_poller) return ERROR("erlang_poller_aldready_registered");

    ctx = enif_alloc(sizeof(struct map_ctx));
    optics_poller_backend(poller, ctx, backend_eo, backend_free_eo);
    erlang_poller = true;
```
This will make things consistent with your `eo_register_carbon_poller` function and also reduces the indentation level of the significant code in your function which tends to make the flow of your function easier to follow.